### PR TITLE
[`BitsAndBytesConfig`] Warning for unused `kwargs` & safety checkers for `load_in_4bit` and `load_in_8bit`

### LINKED ISSUE #29736: BitsAndBytesConfig Accepting Arbitrary Parameters without Throwing Error
### System Info

- `transformers` version: 4.38.2
- Platform: Linux-6.1.58+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.2
- Accelerate version: 0.28.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.2.1+cu121 (True)
- Tensorflow version (GPU?): 2.15.0 (True)
- Flax version (CPU?/GPU?/TPU?): 0.8.2 (cpu)
- Jax version: 0.4.23
- JaxLib version: 0.4.23
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@SunMarc @youn

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. Module Imports
    ```python
    from transformers import BitsAndBytesConfig, AutoModel
    ```

2. Define BitsAndBytesConfig
    ```python
    quant_config = BitsAndBytesConfig({"foo":"bar", "foo":"bar"})

    # also works with:
    # quant_config = BitsAndBytesConfig(**{"foo":"bar", "foo":"bar"})

    quant_config
    """
      >>>
      BitsAndBytesConfig {
      "_load_in_4bit": false,
      "_load_in_8bit": {
        "foo": "bar"
      },
      "bnb_4bit_compute_dtype": "float32",
      "bnb_4bit_quant_type": "fp4",
      "bnb_4bit_use_double_quant": false,
      "llm_int8_enable_fp32_cpu_offload": false,
      "llm_int8_has_fp16_weight": false,
      "llm_int8_skip_modules": null,
      "llm_int8_threshold": 6.0,
      "load_in_4bit": false,
      "load_in_8bit": {
        "foo": "bar"
      },
      "quant_method": "bitsandbytes"
    }
    """
    ```

3. Load in Model with Quant Config
    ```python
    AutoModel.from_pretrained("TinyLlama/TinyLlama-1.1B-Chat-v1.0", 
                               quantization_config=quant_config, 
                               device_map="auto")
    ```

4. Model successfully loads without any error showing user that an invalid `BitsAndBytesConfig` was passed in

### Expected behavior

- It seems like there's no validation for `BitsAndBytesConfig`, resulting in users able to pass in arbitrary parameters without realizing
- Ideally, `BitsAndBytesConfig` should throw and error when invalid config is encountererd.

### PR DESCRIPTION
**Pull request by benjaminye**
# What does this PR do?

Fixes [#29736](https://github.com/huggingface/transformers/issues/29736)
- Introduces a helpful warning when passing in unused keyword arguments when initializing `BitsAndBytesConfig`
- Checks if `load_in_4bit` and `load_in_4bit` is indeed a `boolean` value in order to prevent arbitrary types to be passed in
- Checks for type safety again when setters for `load_in_4bit` and `load_in_8bit` are invoked

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. ([link](https://github.com/huggingface/transformers/issues/29736))
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by amyeroberts — APPROVED** [review]
Additions look good to me - thanks for adding! 

Let's just have a final confirmation from @younesbelkada that it's OK

File: src/transformers/utils/quantization_config.py [hunk 1]
```diff
@@ -296,6 +302,9 @@ def load_in_8bit(self):
 
     @load_in_8bit.setter
     def load_in_8bit(self, value: bool):
+        if not isinstance(value, bool):
+            raise ValueError("load_in_4bit must be a boolean")
```
Comment (younesbelkada) [inline src/transformers/utils/quantization_config.py:306]:
```suggestion
            raise ValueError("load_in_8bit must be a boolean")
```

**Review by younesbelkada — APPROVED** [review]
Looking great thanks !

**Comment by benjaminye** [comment]
@amyeroberts @younesbelkada
Thanks everyone! 😃

**Pull request merged**